### PR TITLE
Avoid making invalid "Use let" suggestions in ApplicativeDo blocks

### PR DESCRIPTION
**Pull request by zliu41**
In an ApplicativeDo block, the second last statement cannot be a `let`, for example, this doesn't compile:
```haskell
{-# LANGUAGE ApplicativeDo, DeriveFunctor #-}

newtype T a = T a deriving Functor

instance Applicative T where
  pure = T
  T a <*> T b = T (a b)

foo :: T Int
foo = do
  x <- pure 2
  let y = 3
  pure $ x + y
```

So we avoid making such suggestions on the second last statement, if the do-block is possibly an ApplicativeDo block.

### CODE REVIEW

**Comment by ndmitchell** [comment]
Should this be reported upstream to GHC? It seems a pretty arbitrary limitation...

File: src/Hint/Monad.hs [hunk 1]
```diff
@@ -238,8 +239,21 @@ monadSteps _ _ = []
...
+    go f = \case
+      [] -> []
+      -- If the do block is possibly an ApplicativeDo block, don't make a suggestion on the
+      -- second last statement.
+      [_, x] | mayBeApplicativeDo x -> go f [x]
+      x : xs -> maybe id (:) (f x) (go f xs)
+
+    -- A do block is possibly an ApplicativeDo block if the last statement is one of
+    -- 'pure ...' , 'pure $ ...', 'return ...' or 'return $ ...'.
+    mayBeApplicativeDo = \case
```
Comment (googleson78) [inline src/Hint/Monad.hs:253]: I think this will have too many false positives. I would wager *most* do blocks end like this :/

Comment (ndmitchell) [reply src/Hint/Monad.hs:253]: Agreed - but at the same time - the hint isn't the greatest in the world, so I'm less concerned than I would be otherwise.

Comment (googleson78) [reply src/Hint/Monad.hs:253]: I think I've seen it a few times in student code, so I do consider it to be useful as a teaching tool at least.

**Comment by zliu41** [comment]
Yeah, agreed it should be reported upstream first to see if it's a GHC bug.

**Pull request closed without merging**